source/Nevermore/IQueryBuilder.cs

```diff
@@ -1,40 +1,85 @@
 using System;
 using System.Collections.Generic;
-using Nevermore.Joins;
+using Nevermore.AST;
 
 namespace Nevermore
 {
-    public interface IQueryBuilder<TRecord> where TRecord : class
+    public interface IOrderedQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+    }
+
+    public interface ITableSourceQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
     {
-        IQueryGenerator QueryGenerator { get; }
+        ITableSourceQueryBuilder<TRecord> View(string viewName);
+        ITableSourceQueryBuilder<TRecord> Table(string tableName);
+        ITableSourceQueryBuilder<TRecord> Alias(string tableAlias); 
+        IQueryBuilder<TRecord> Hint(string tableHint);
+        IAliasedSelectSource AsAliasedSource();
+    }
 
+    public interface ISubquerySourceBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+        ISubquerySource AsSource();
+        ISubquerySourceBuilder<TRecord> Alias(string subqueryAlias);
+    }
+
+    public interface IJoinSourceQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+        IJoinSourceQueryBuilder<TRecord> On(string leftField, JoinOperand operand, string rightField);
+    }
+
+    public interface IQueryBuilder<TRecord> where TRecord : class
```

Comment:
How hard would it be to add an immutable QueryBuilder?

Reply:
I think that would be very easy :)

Reply:
Should we add it from the get go?

Reply:
Personally I would rather add it later to limit the scope of work. It is a non breaking change so can be in a minor version release. But open to being convinced otherwise.